tests/test_scenarios.py

```diff
@@ -2,26 +2,129 @@
 from textwrap import indent
 from typing import Any, Dict
 
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric.rsa import (
+        generate_private_key, RSAPrivateKey)
+
 from proof_of_concept.components.ddm_site import Site
+from proof_of_concept.components.registry_client import RegistryClient
 from proof_of_concept.definitions.assets import ComputeAsset, DataAsset
+from proof_of_concept.definitions.registry import (
+        PartyDescription, SiteDescription)
 from proof_of_concept.definitions.workflows import Job, WorkflowStep, Workflow
 from proof_of_concept.policy.rules import (
     InAssetCollection, MayAccess, ResultOfDataIn,
     ResultOfComputeIn)
+from proof_of_concept.rest.ddm_site import SiteRestApi, SiteServer
 
 
 logger = logging.getLogger(__file__)
 
 
+def create_parties(
+        site_descriptions: Dict[str, Any]
+        ) -> Dict[str, RSAPrivateKey]:
+    """Creates parties with private keys and registers them."""
+    return {
+            desc['owner']: generate_private_key(
+                public_exponent=65537,
+                key_size=2048,
+                backend=default_backend())
+            for desc in site_descriptions.values()}
+
+
+def register_parties(
+        registry_client: RegistryClient, parties: Dict[str, RSAPrivateKey]
+        ) -> None:
+    """Register parties with their public keys."""
+    for party_id, private_key in parties.items():
+        registry_client.register_party(
+                PartyDescription(party_id, private_key.public_key()))
+
+
+def sign_rules(
+        site_descriptions: Dict[str, Any], parties: Dict[str, RSAPrivateKey]
+        ) -> None:
+    """Update site descriptions by signing rules."""
+    for desc in site_descriptions.values():
+        private_key = parties[desc['owner']]
+        for rule in desc['rules']:
+            rule.sign(private_key)
+
+
+def create_sites(
+        registry_client: RegistryClient,
+        site_descriptions: Dict[str, Any]
+        ) -> Dict[str, Site]:
+    """Creates sites for the scenario."""
+    return {
+            site_name: Site(
+                site_name, desc['owner'], desc['namespace'], desc['assets'],
+                desc['rules'], registry_client)
+            for site_name, desc in site_descriptions.items()}
+
+
+def create_servers(sites: Dict[str, Site]) -> Dict[str, SiteServer]:
+    """Create REST servers for sites."""
+    return {
+            site_name: SiteServer(
+                    SiteRestApi(site.policy_store, site.store, site.runner))
+            for site_name, site in sites.items()}
```

Comment:
If I understand it correctly this is now different code than what a user would use (then you use `wsgi_app` based on settings right?). And I believe I remember from the RegistryServer code that you somehow reused more of the code to start the server in the tests. Maybe you could try something like that here?

Reply:
Well, this test uses the prototype in everything-in-one-process mode, while the `wsgi_app` is for use in distributed mode. 

There will eventually be a similar driver for the distributed set-up, so there will be a docker-compose (or ansible or something, if you're running across multiple machines) there that starts up a registry and a couple of sites, and then some containers that represent users which connect to a site and submit assets and workflows and policies. But I need to build a few more REST APIs before that can work.

So yeah, maybe eventually this will get turned into an experiment definition support library for the local running case, but for now it just needs to run these tests.

Reply:
Cool!